File: src/main/java/com/tuuzed/feedparser/AtomParser.java
/* Copyright 2017 TuuZed
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tuuzed.feedparser;


import com.tuuzed.feedparser.util.DateUtils;
import org.xmlpull.v1.XmlPullParser;

import java.util.Map;

class AtomParser extends GenericParser {
    private boolean isBeginEntry = false;
    private boolean isBeginEntryAuthor = false;
    private boolean isBeginEntryContributor = false;
    private String[] tempStrArr = new String[3];

    @Override
    public void startTag(String tagName, XmlPullParser xmlPullParser, FeedHandler handler) {
        switch (tagName) {
            case "feed":
                handler.begin();
                break;
            case "entry":
                isBeginEntry = true;
                handler.entryBegin();
                break;
        }
        if (isBeginEntry) {
            entry(tagName, xmlPullParser, handler);
        } else {
            feed(tagName, xmlPullParser, handler);
        }
    }

    private void feed(String tagName, XmlPullParser xmlPullParser, FeedHandler handler) {
        switch (tagName) {
            case "title": {
                String text = nextText(xmlPullParser);
                handler.title(text);
                break;
            }
            case "subtitle": {
                String text = nextText(xmlPullParser);
                handler.subtitle(text);
                break;
            }
            case "link":
                Map<String, String> attrs = getAttrs(xmlPullParser);
                if (attrs != null) {
                    handler.link(attrs.get("type"), attrs.get("href"), attrs.get("title"));
                } else {
                    handler.link(null, null, null);
                }
                break;
        }
    }

    private void entry(String tagName, XmlPullParser xmlPullParser, FeedHandler handler) {
        switch (tagName) {
            case "author": {
                isBeginEntryAuthor = true;
                break;
            }
            case "name": {
                if (isBeginEntryAuthor || isBeginEntryContributor) {
                    tempStrArr[0] = nextText(xmlPullParser);
                }
                break;
            }
            case "uri": {
                if (isBeginEntryAuthor || isBeginEntryContributor) {
                    tempStrArr[1] = nextText(xmlPullParser);
                }
                break;
            }
            case "email": {
                if (isBeginEntryAuthor || isBeginEntryContributor) {
                    tempStrArr[1] = nextText(xmlPullParser);
                }
                break;
            }
            case "content": {
                Map<String, String> attrs = getAttrs(xmlPullParser);
                String text = nextText(xmlPullParser);
                if (attrs != null) {
                    handler.entryContent(attrs.get("type"), attrs.get("language"), text);
                } else {
                    handler.entryContent(null, null, text);
                }
                break;
            }
            case "contributor": {
                isBeginEntryContributor = true;
                break;
            }
            case "id": {
                String text = nextText(xmlPullParser);
                handler.entryId(text);
                break;
            }
            case "link": {
                Map<String, String> attrs = getAttrs(xmlPullParser);
                if (attrs != null) {
                    handler.entryLink(attrs.get("type"), attrs.get("href"), attrs.get("title"));
                } else {
                    handler.entryLink(null, null, null);
                }
                break;
            }
            case "published": {
                String text = nextText(xmlPullParser);
                handler.entryPublished(DateUtils.parse(text), text);
                break;
            }
            case "summary": {
                Map<String, String> attrs = getAttrs(xmlPullParser);
                String text = nextText(xmlPullParser);
                if (attrs != null) {
                    handler.entrySummary(attrs.get("type"), attrs.get("language"), text);
                } else {
                    handler.entrySummary(null, null, text);
                }
                break;
            }
            case "category": {
                Map<String, String> attrs = getAttrs(xmlPullParser);
                String text = nextText(xmlPullParser);
                if (attrs != null) {
                    handler.entryTags(attrs.get("term"), attrs.get("scheme"), text);
                } else {
                    handler.entryTags(null, null, text);
                }
                break;
            }
            case "title": {
                String text = nextText(xmlPullParser);
                handler.entryTitle(text);
                break;
            }
            case "updated": {
                String text = nextText(xmlPullParser);
                handler.entryUpdated(DateUtils.parse(text), text);
                break;
            }
        }
    }

    @Override
    public void endTag(String tagName, FeedHandler handler) {
        switch (tagName) {
            case "feed":
                handler.end();
                break;
            case "entry":
                isBeginEntry = false;
                handler.entryEnd();
                break;
            case "author":
                handler.entryAuthor(tempStrArr[0], tempStrArr[1], tempStrArr[2]);
                emptyTempStrArr();
                isBeginEntryAuthor = false;
                break;
            case "contributor":
                handler.entryContributor(tempStrArr[0], tempStrArr[1], tempStrArr[2]);
                emptyTempStrArr();
                isBeginEntryContributor = false;
                break;
        }
    }

    private void emptyTempStrArr() {
        for (int i = 0; i < tempStrArr.length; i++) {
            tempStrArr[i] = null;
        }
    }
}

File: src/main/kotlin/com/tuuzed/feedparser/AtomParser.kt
package com.tuuzed.feedparser

import com.tuuzed.feedparser.ext.clear
import com.tuuzed.feedparser.ext.getAttrs
import com.tuuzed.feedparser.ext.getNextText
import com.tuuzed.feedparser.util.DateUtils
import org.xmlpull.v1.XmlPullParser

internal class AtomParser : GenericParser() {
    private var isBeginEntry = false
    private var isBeginEntryAuthor = false
    private var isBeginEntryContributor = false
    private val tempStrArr = arrayOfNulls<String>(3)

    override fun startTag(tagName: String, pullParser: XmlPullParser, handler: FeedHandler) {
        when (tagName) {
            "feed" -> handler.begin()
            "entry" -> {
                isBeginEntry = true
                handler.entryBegin()
            }
        }
        if (isBeginEntry) {
            entry(tagName, pullParser, handler)
        } else {
            feed(tagName, pullParser, handler)
        }
    }

    private fun feed(tagName: String, xmlPullParser: XmlPullParser, handler: FeedHandler) {
        when (tagName) {
            "title" -> handler.title(xmlPullParser.getNextText())
            "subtitle" -> handler.subtitle(xmlPullParser.getNextText())
            "link" -> {
                val attrs = xmlPullParser.getAttrs()
                if (attrs != null) {
                    handler.link(attrs["type"], attrs["href"], attrs["title"])
                } else {
                    handler.link(null, null, null)
                }
            }
        }
    }

    private fun entry(tagName: String, xmlPullParser: XmlPullParser, handler: FeedHandler) {
        when (tagName) {
            "author" -> isBeginEntryAuthor = true
            "name" -> if (isBeginEntryAuthor || isBeginEntryContributor) {
                tempStrArr[0] = xmlPullParser.getNextText()
            }
            "uri" -> if (isBeginEntryAuthor || isBeginEntryContributor) {
                tempStrArr[1] = xmlPullParser.getNextText()
            }
            "email" -> if (isBeginEntryAuthor || isBeginEntryContributor) {
                tempStrArr[1] = xmlPullParser.getNextText()
            }
            "content" -> {
                val attrs = xmlPullParser.getAttrs()
                if (attrs != null) {
                    handler.entryContent(attrs["type"], attrs["language"], xmlPullParser.getNextText())
                } else {
                    handler.entryContent(null, null, xmlPullParser.getNextText())
                }
            }
            "contributor" -> isBeginEntryContributor = true
            "id" -> handler.entryId(xmlPullParser.getNextText())
            "link" -> {
                val attrs = xmlPullParser.getAttrs()
                if (attrs != null) {
                    handler.entryLink(attrs["type"], attrs["href"], attrs["title"])
                } else {
                    handler.entryLink(null, null, null)
                }
            }
            "published" -> {
                val text = xmlPullParser.getNextText()
                handler.entryPublished(DateUtils.parse(text), text)
            }
            "summary" -> {
                val attrs = xmlPullParser.getAttrs()
                val text = xmlPullParser.getNextText()
                if (attrs != null) {
                    handler.entrySummary(attrs["type"], attrs["language"], text)
                } else {
                    handler.entrySummary(null, null, text)
                }
            }
            "category" -> {
                val attrs = xmlPullParser.getAttrs()
                val text = xmlPullParser.getNextText()
                if (attrs != null) {
                    handler.entryTags(attrs["term"], attrs["scheme"], text)
                } else {
                    handler.entryTags(null, null, text)
                }
            }
            "title" -> {
                val text = xmlPullParser.getNextText()
                handler.entryTitle(text!!)
            }
            "updated" -> {
                val text = xmlPullParser.getNextText()
                handler.entryUpdated(DateUtils.parse(text)!!, text!!)
            }
        }
    }

    override fun endTag(tagName: String, handler: FeedHandler) {
        when (tagName) {
            "feed" -> handler.end()
            "entry" -> {
                isBeginEntry = false
                handler.entryEnd()
            }
            "author" -> {
                handler.entryAuthor(tempStrArr[0], tempStrArr[1], tempStrArr[2])
                tempStrArr.clear()
                isBeginEntryAuthor = false
            }
            "contributor" -> {
                handler.entryContributor(tempStrArr[0], tempStrArr[1], tempStrArr[2])
                tempStrArr.clear()
                isBeginEntryContributor = false
            }
        }
    }
}